### PR TITLE
Make global AD indexing the default config

### LINKED ISSUE #16091: Make global AD indexing with sparse container MOOSE default config
## Reason
Pros:
- Much simpler code
- Much lower chance of accidentally creating slow code because if you try to do a bunch of `DualNumber` math with an unseeded sparse container, you don't take much of a performance hit. You can see #16089 for an example of why nonsparse containers are a pain
- Much more flexible code. You can build residuals depending on totally arbitrary nonlinear degrees of freedom and track it with ease
- Maintain speed over a broad class of physics and dimensionality. It's nice to able to do a 3D Navier-Stokes calculation and a 1D diffusion calculation with the same binary and not see a penalty in performance in the 1D calculation

Cons:
- Maybe it's slower sometimes? Someone has yet to show me a concrete example of this, however.

## Design
It's already done. Just have to change `MooseDefaultConfig.h`. I'm going to give users some time to give feedback before making this change. If I see enough concrete examples of global-sparse AD being slow relative to local-nonsparse AD, then I won't make the change. ***Unless that happens, however, I'm planning to make the change in January.***

An important note is that even if we change to global-sparse as the default config, nonsparse isn't going to disappear as a configure option, at least not for the foreseeable future and not without proper deprecation warnings (they would be compiler-time warnings).

## Impact
I think I would probably spend a lot less time doing AD maintenance in a framework that was global-sparse-AD based. For example I've sunk a half-day into #16089 which I wouldn't have if we lived in a world of global-sparse-AD. Also if we were to one day deprecate the nonsparse code, our AD code base would get much simpler and it would be easier to do AD development. Moreover, I think it would positively impact users for the reasons mentioned above.

### PR DESCRIPTION
**Pull request by lindsayad**
Closes #16091

I'm tired of forgetting to run configure for INSFV and nobody has shown me any profiles/reasons not to make the switch.

### CODE REVIEW

**Comment by loganharbour** [comment]
We need to change the sparse test in CIVET to use non-global once this is good.

File: configure.ac [hunk 1]
```diff
@@ -50,19 +50,15 @@ AC_MSG_RESULT([configuring with derivative backing array size of $derivative_siz
...
         [
           AS_IF([test "$derivative_type" = nonsparse],
                  [
-                   AS_IF([test "$user_set_derivative_type" = yes],
-                         dnl The user explicitly requested a nonsparse container which is incompatabile with global indexing
-                         [AC_MSG_ERROR(cannot request global AD indexing and a nonsparse derivative container)],
-                         dnl The user did not explicity request a container type so well set sparse for them
-                         [derivative_type=sparse]
-                        )
+                   dnl The user explicitly requested a nonsparse container which is incompatabile with global indexing
```
Comment (GiudGiud) [inline configure.ac:60]:
```suggestion
                   dnl The user explicitly requested a non-sparse container which is incompatible with global indexing
```

File: configure [hunk 2]
```diff
@@ -1763,12 +1763,7 @@ case "$ad_indexing_type" in #(
 
           if test "$derivative_type" = nonsparse; then :
 
-                   if test "$user_set_derivative_type" = yes; then :
-                           as_fn_error $? "cannot request global AD indexing and a nonsparse derivative container" "$LINENO" 5
-else
-                           derivative_type=sparse
-
-fi
+                                      as_fn_error $? "cannot use global AD indexing with a nonsparse derivative container" "$LINENO" 5
```
Comment (GiudGiud) [inline configure:1766]:
```suggestion
                                      as_fn_error $? "cannot use global AD indexing with a non-sparse derivative container" "$LINENO" 5
```

Comment (GiudGiud) [reply configure:1766]: indenting seems wrong?

**Review by GiudGiud — COMMENTED** [review]
You had said January!
nah jk I agree with this change.

**Comment by lindsayad** [comment]
configure is generated automatically by autotools so we do not want to modify it

**Comment by lindsayad** [comment]
Oh the suckeye is going to need #16397 in first

**Comment by tophmatthews** [comment]
This makes sparse default too right...do it!

**Comment by lindsayad** [comment]
I like that user feedback!

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
So before this PR, what is being cached?
Also what do you mean by 'Jacobian mortar' ?

**Review by loganharbour — APPROVED** [review]

**Comment by lindsayad** [comment]
> So before this PR, what is being cached?

Before this PR, we were not caching any Jacobians during mortar constraint evaluation when using the local AD indexing scheme. Instead we were adding straight into the matrix right then and there. However, we *were* doing caching for the gobal AD indexing scheme. Now *after* this PR we cache for both types of indexing. What we cache: all parts of the Jacobian arising from mortar constraint evaluation. I think I actually have pretty good doxygen for this:
```c++
  /**
   * Cache *all* portions of the Jacobian, e.g. LowerLower, LowerSecondary, LowerPrimary,
   * SecondaryLower, SecondarySecondary, SecondaryPrimary, PrimaryLower, PrimarySecondary,
   * PrimaryPrimary for mortar-like objects. Primary indicates the interior parent element on the
   * primary side of the mortar interface. Secondary indicates the interior parent element on the
   * secondary side of the interface. Lower denotes the lower-dimensional element living on the
   * secondary side of the mortar interface; it's the boundary face of the \p Secondary element.
   */
```
> Also what do you mean by 'Jacobian mortar' ? It's the Jacobian contribution from the elements that are part of thermal / mechanical constraints mortar?

Hopefully the above answers this. It applies to anything that derives from `MortarConstraintBase`. It can be thermal or mechanical contact or it can be a constraint that enforces solution continuity across a gap (or even no-gap, especially useful for non-conforming meshes).

**Pull request merged**